src/treeitem_dependency.ml

```diff
@@ -0,0 +1,180 @@
+
+let of_string s = s |> Sexp.of_string |> t_of_sexp
+
+let label = function
+  | Sandbox (Opam (_, Named name)) -> name
+  | Sandbox (Opam (_, Local path)) ->
+    let name = Path.to_string path in
+    name
+  | Sandbox _ ->
+    (* TODO: Support other sandboxes *)
+    assert false
+  | Package (dep, _) -> Sandbox.Package.name dep
+
+let description = function
+  | Sandbox (Opam (opam, switch)) -> Opam.switch_compiler opam switch
+  | Sandbox _ ->
+    (* TODO: Support other sandboxes *)
+    assert false
+  | Package (dep, _) -> Promise.return (Some (Sandbox.Package.version dep))
+
+let tooltip = function
+  | Sandbox (Opam (_, Named _)) -> None
+  | Sandbox (Opam (_, Local _)) -> None
+  | Sandbox _ ->
+    (* TODO: Support other sandboxes *)
+    assert false
+  | Package (dep, _) -> Sandbox.Package.synopsis dep
+
+let context_value = function
+  | Sandbox (Opam _) -> "opam-switch"
+  | Sandbox _ ->
+    (* TODO: Support other sandboxes *)
+    assert false
+  | Package (dep, _) -> (
+    match Sandbox.Package.documentation dep with
+    | Some _ -> "package-with-doc"
+    | None -> "package" )
+
+let icon ~extension_path = function
+  | Sandbox _ ->
+    TreeItem.LightDarkIcon.
+      { light = `String (extension_path ^ "/assets/dependency-light.svg")
+      ; dark = `String (extension_path ^ "/assets/dependency-dark.svg")
+      }
+  | Package _ ->
+    TreeItem.LightDarkIcon.
+      { light = `String (extension_path ^ "/assets/number-light.svg")
+      ; dark = `String (extension_path ^ "/assets/number-dark.svg")
+      }
+
+let collapsible_state = function
+  | Sandbox _ -> Vscode.TreeItemCollapsibleState.Collapsed
+  | Package (dep, _) -> (
+    match Sandbox.Package.depends dep with
+    | Some []
+    | None ->
+      TreeItemCollapsibleState.None
+    | _ -> TreeItemCollapsibleState.Collapsed )
+
+let to_treeitem ~extension_path dependency =
+  let open Promise.Syntax in
+  let icon = `LightDark (icon ~extension_path dependency) in
+  let collapsibleState = collapsible_state dependency in
+  let label = Vscode.TreeItemLabel.create ~label:(label dependency) () in
+  let item = Vscode.TreeItem.make ~label ~collapsibleState () in
+  TreeItem.set_id item (to_string dependency);
+  TreeItem.set_iconPath item icon;
+  TreeItem.set_contextValue item (context_value dependency);
+  let+ _ =
+    Promise.Option.iter
+      (fun desc -> TreeItem.set_description item (`String desc))
+      (description dependency)
+  in
+  Option.iter (tooltip dependency) ~f:(fun desc ->
+      TreeItem.set_tooltip item (`String desc));
+  item
+
+let get_dependencies =
+  let open Promise.Syntax in
+  function
+  | Sandbox sandbox -> (
+    let* packages = Sandbox.root_packages sandbox in
+    match packages with
+    | Ok packages ->
+      let names =
+        packages |> List.map ~f:(fun n -> Package (n, Some (Sandbox sandbox)))
+      in
+      Promise.return (Some names)
+    | Error err ->
+      let* _ =
+        Vscode.Window.showInformationMessage
```

Comment:
We have `show_message` to simplify code like this.

Reply:
Done